### PR TITLE
Restrict nuget version range allows incorrect ranges

### LINKED ISSUE #9145: NuGetVersionRange allows logically incorrect ranges to be parsed
See: https://nugettoolsdev.azurewebsites.net/5.5.0-preview.2.6382/parse-version-range?versionRange=%5B10.0.0%2C+9.0.0%29

> [10.0.0, 9.0.0)

This range is empty and there's no reasonable scenario where it's necessary. 

This has always been the case, all the way to 2.5 https://nugettoolsdev.azurewebsites.net/2.5.0/parse-version-range?versionRange=%5B10.0.0%2C+9.0.0%29.

### PR DESCRIPTION
**Pull request by erdembayar**
Fixes: https://github.com/NuGet/Home/issues/9145

Regression? Last working version:

## Description
See: https://nugettoolsdev.azurewebsites.net/5.5.0-preview.2.6382/parse-version-range?versionRange=%5B10.0.0%2C+9.0.0%29

> [10.0.0, 9.0.0)

This range is empty and there's no reasonable scenario where it's necessary. 

This has always been the case, all the way to 2.5 https://nugettoolsdev.azurewebsites.net/2.5.0/parse-version-range?versionRange=%5B10.0.0%2C+9.0.0%29.

Now we'll throw exception in above case. Also we consider [9.0.0, 9.0.0), (9.0.0, 9.0.0] as illogical now.
But we're not touch (9.0.0, 9.0.0) even though it's technically illogical range, it's used as empty range in many unit test cases, let's not overdo it.
## PR Checklist

- [x] PR has a meaningful title
- [ ] PR has a linked issue.
- [ ] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by erdembayar** [comment]
@nkolev92 
How about "[1.0.0, 1.0.0)"? Some tests are failing.
Personally, I want to treat as bad range.

**Comment by nkolev92** [comment]
That seems like a bad range, but I'm not sure what you are to tell me by "some tests are failing".

**Comment by erdembayar** [comment]
> That seems like a bad range, but I'm not sure what you are to tell me by "some tests are failing".

For example below test is now failing, because input is actually [2.3.1-RC+srv01-a5c5ff9, 2.3.1-RC+srv02-dbf5ec0) means [2.3.1-RC, 2.3.1-RC) ~ [a,a) is here. I believe [a,a) is illogical range. Something can't be exluding and including same thing `a` same time.

https://github.com/NuGet/NuGet.Client/blob/ea6602cdf93f725e9bdc1682b69d55c751b1fd6a/test/NuGet.Core.Tests/NuGet.Versioning.Test/ExternalComparerTests.cs#L15-L19

**Comment by nkolev92** [comment]
I think that the test could be better.

**Comment by erdembayar** [comment]
> I think that the test could be better.

Removed only that case, it's illogical test input.

**Comment by erdembayar** [comment]
@nkolev92 
`[2.0.0)` it it illogical range? I guess it's. I believe it should be `[2.0.0, )` or `[2.0.0, *)` valid, yeah?

**Comment by nkolev92** [comment]
`[2.0.0)` likely normalizes to `[2.0.0,2.0.0)` which is equivalent to the one you mentioned above.
`[2.0.0, )` is different and it's the normalized version of `2.0.0`.
 
 `[2.0.0, *)` is not a valid range, because the floating part cannot be part of the upper range. Only a min range can be floating.

**Comment by erdembayar** [comment]
@NuGet/nuget-client 
Hi.
Ready for review, please review.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs [hunk 1]
```diff
@@ -1503,5 +1509,29 @@ public void FindBestMatch_FloatingPrereleaseMinor_WithExtraDashes()
...
+        [InlineData("[*, *)")]
+        [InlineData("(*, *]")]
+        [InlineData("[1.0.0-beta, 1.0.0-beta+900)")]
+        [InlineData("(1.0.0-beta+600, 1.0.0-beta]")]
+        [InlineData("(1.0)")]
+        [InlineData("(1.0.0)")]
+        [InlineData("[2.0.0)")]
+        [InlineData("(2.0.0]")]
+        [InlineData("[2.0.0, *)")]
+        public void VersionRange_LogicallyIncorrectRanges_Throws(string range)
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs]: The correct name according to the guide would be `Parse_****_****`

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs [hunk 2]
```diff
@@ -1503,5 +1509,29 @@ public void FindBestMatch_FloatingPrereleaseMinor_WithExtraDashes()
...
+
+        [Theory]
+        [InlineData("[1.1.4, 1.1.2)")]
+        [InlineData("[1.1.4, 1.1.2]")]
+        [InlineData("(1.1.4, 1.1.2)")]
+        [InlineData("(1.1.4, 1.1.2]")]
+        [InlineData("[1.0.0, 1.0.0)")]
+        [InlineData("(1.0.0, 1.0.0]")]
+        [InlineData("(1.0, 1.0.0]")]
+        [InlineData("[*, *)")]
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs]: I sincerely hope that there are tests for this already :) `*` should not be allowed in the upper version.

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs [hunk 3]
```diff
@@ -233,6 +236,19 @@ public static bool TryParse(string value, bool allowFloating, out VersionRange v
                 }
             }
 
+            if (minVersion != null && maxVersion != null
+                && minVersion >= maxVersion
+                && !(minVersion == maxVersion && isMinInclusive && isMaxInclusive))
+            {
+                if ((partsLength == 1 && minVersion == maxVersion
+                            && (isMinInclusive ^ isMaxInclusive || !isMinInclusive && !isMaxInclusive))
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: I'm having a hard time reading this statement. 

Can you try doing another pass on simplifying this logic please? 

For example: `isMinInclusive ^ isMaxInclusive || !isMinInclusive && !isMaxInclusive` can be rewritten as `!(isMinInclusive && isMaxInclusive)` 

Given the complexity of the checks, I think some comments would be helpful.

**Review by nkolev92 — COMMENTED** [review]
The general idea seems good. 
Maybe we can try simplifying the code a bit? 

Why does your checklist have tests exception selected?

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs (hunk 2 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs]: Removed.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs (hunk 1 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs]: Ok . Renamed.

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs (hunk 3 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Ok. I simplified the logic and made it easy to reason with small parts. Please check.

**Comment by erdembayar** [comment]
> The general idea seems good.
> Maybe we can try simplifying the code a bit?
> 
> Why does your checklist have tests exception selected?

Thank you for your review. I addressed your comments and made it more easy to reason. Please review again.

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs [hunk 4]
```diff
@@ -233,6 +236,32 @@ public static bool TryParse(string value, bool allowFloating, out VersionRange v
                 }
             }
 
+            // Illogical version range detection
+            if (minVersion != null && maxVersion != null
+                && minVersion >= maxVersion
+                && !(minVersion == maxVersion && isMinInclusive && isMaxInclusive))  // Exclude (9.0.0,9.0.0) since it's used as empty version range.
+            {
+                if (partsLength == 1)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: This check can still be a lot simpler. 

All comments aside, I'd expect `return false` covers all scenarios here and you don't need any of the parts checks.

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs [hunk 5]
```diff
@@ -233,6 +236,32 @@ public static bool TryParse(string value, bool allowFloating, out VersionRange v
...
+                {
+                    var onePartValidation = isMinInclusive ^ isMaxInclusive; // (1.0.0] and [1.0.0) are invalid
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: As mentioned above, you can simplify this by writing it as

> !(isMinInclusive && isMaxInclusive)

You can check this with a truth table or a quick console app. 

```cs
static void Main(string[] args)
        {
            Compare(true, true);
            Compare(true, false);
            Compare(false, true);
            Compare(false, false);

        }

        public static void Compare(bool isMinInclusive, bool isMaxInclusive)
        {
            if (Calculate(isMinInclusive, isMaxInclusive) != CalculateSimple(isMinInclusive, isMaxInclusive))
            {
                Console.WriteLine(isMinInclusive + " " + isMaxInclusive + " Original: " + Calculate(isMinInclusive, isMaxInclusive) + " Updated: " + CalculateSimple(isMinInclusive, isMaxInclusive));

            }
        }

        private static bool CalculateSimple(bool isMinInclusive, bool isMaxInclusive)
        {
            return !(isMinInclusive && isMaxInclusive);
        }

        public static bool Calculate(bool isMinInclusive, bool isMaxInclusive)
        {
            var onePartValidation = isMinInclusive ^ isMaxInclusive;
            onePartValidation |= !isMinInclusive && !isMaxInclusive;
            return onePartValidation;

        }
```

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs [hunk 6]
```diff
@@ -233,6 +236,32 @@ public static bool TryParse(string value, bool allowFloating, out VersionRange v
...
+                    onePartValidation |= !isMinInclusive && !isMaxInclusive; ; // or (1.0.0) is invalid
+                    if (onePartValidation)
+                    {
+                        return false;
+                    }
+                }
+                else
+                {
+                    var manyPartsValidation = minVersion > maxVersion // If a > b then (a, b) range is invalid. But below 2 cases tests are a = b (equals).
+                                              | isMinInclusive // [1.0.0, 1.0.0) is invalid
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: This check is pretty much equivalent to the above one. 

I'd take a look at [De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws).

**Review by nkolev92 — COMMENTED** [review]
I think we can go even simpler than what we have right now :)

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs [hunk 7]
```diff
@@ -233,6 +236,32 @@ public static bool TryParse(string value, bool allowFloating, out VersionRange v
...
+                                              | isMaxInclusive; // (1.0.0, 1.0.0] and [1.0.0, 1.0.0] are invalid
+                    if (manyPartsValidation)
+                    {
+                        return false;
+                    }
+                }
+            }
+
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: I think we can change the above logic as follows. Hope it covers all edge cases :)
- Add below logic after https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs#L187
```cs
                //(1.0.0) , (1.0.0] and [1.0.0) invalid [1.0.0] is the only valid version number
                if (parts.Length == 1
                    && !(isMinInclusive && isMaxInclusive))
                {
                    return false;
                }
```
- Add below logic after https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs#L235
```cs
            if (minVersion != null && maxVersion != null)
            {
                int result = minVersion.CompareTo(maxVersion);

                //minVersion is equal to maxVersion (1.0.0, 1.0.0], [1.0.0, 1.0.0)
                if (result == 0
                    && (isMinInclusive ^ isMaxInclusive))
                    return false;

                //minVersion > maxVersion
                if (result > 0)
                    return false;
            }
```

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs (hunk 4 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Ok. I addressed this one. Please check.

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs (hunk 5 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Taken your suggestion.

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs (hunk 7 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Ok. Taken your suggestion. Please review again now.

**Comment by erdembayar** [comment]
Addressed PR review comments. Ready for review.

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs [hunk 8]
```diff
@@ -123,6 +123,7 @@ public static bool TryParse(string value, bool allowFloating, out VersionRange v
             NuGetVersion minVersion = null;
             NuGetVersion maxVersion = null;
             FloatRange floatRange = null;
+            int partsLength = 0;
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: you probably don't need this here anymore. Everything is in the internal `{ }`

**Review by nkolev92 — APPROVED** [review]
1 clean-up nit. 

Nicely done! 👏

**Comment by erdembayar** [comment]
@kartheekp-ms 
Can review again please?

**Review by kartheekp-ms — APPROVED** [review]

**Pull request merged**